File: Zig/priority_queue_utils/src/priority_queue.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const Order = std.math.Order;

/// Priority Queue implementation using a binary heap
/// Supports both min-heap and max-heap configurations
pub fn PriorityQueue(comptime T: type) type {
    return struct {
        const Self = @This();

        items: []T,
        len: usize,
        capacity: usize,
        allocator: Allocator,
        compare: *const fn (a: T, b: T) Order,

        /// Initialize a new priority queue
        pub fn init(allocator: Allocator, compare: *const fn (a: T, b: T) Order) Self {
            return .{
                .items = &[_]T{},
                .len = 0,
                .capacity = 0,
                .allocator = allocator,
                .compare = compare,
            };
        }

        /// Initialize with pre-allocated capacity
        pub fn initCapacity(allocator: Allocator, capacity: usize, compare: *const fn (a: T, b: T) Order) Allocator.Error!Self {
            const items = try allocator.alloc(T, capacity);
            return .{
                .items = items,
                .len = 0,
                .capacity = capacity,
                .allocator = allocator,
                .compare = compare,
            };
        }

        /// Free all allocated memory
        pub fn deinit(self: *Self) void {
            if (self.capacity > 0) {
                self.allocator.free(self.items);
            }
            self.* = undefined;
        }

        /// Check if the queue is empty
        pub fn isEmpty(self: Self) bool {
            return self.len == 0;
        }

        /// Get the number of elements
        pub fn count(self: Self) usize {
            return self.len;
        }

        /// Peek at the top element without removing it
        pub fn peek(self: Self) ?T {
            if (self.isEmpty()) return null;
            return self.items[0];
        }

        /// Insert a new element
        pub fn insert(self: *Self, item: T) Allocator.Error!void {
            try self.ensureCapacity(self.len + 1);
            self.items[self.len] = item;
            self.len += 1;
            self.siftUp(self.len - 1);
        }

        /// Remove and return the top element
        pub fn remove(self: *Self) ?T {
            if (self.isEmpty()) return null;
            
            const top = self.items[0];
            self.len -= 1;
            
            if (self.len > 0) {
                self.items[0] = self.items[self.len];
                self.siftDown(0);
            }
            
            return top;
        }

        /// Ensure capacity is at least new_capacity
        pub fn ensureCapacity(self: *Self, new_capacity: usize) Allocator.Error!void {
            if (new_capacity <= self.capacity) return;
            
            const better_capacity = @max(self.capacity * 2, new_capacity);
            const new_items = try self.allocator.realloc(self.items, better_capacity);
            self.items = new_items;
            self.capacity = better_capacity;
        }

        /// Clear all elements (does not free memory)
        pub fn clear(self: *Self) void {
            self.len = 0;
        }

        /// Convert to a sorted slice (caller owns the memory)
        pub fn toSortedSlice(self: *Self, allocator: Allocator) Allocator.Error![]T {
            const result = try allocator.alloc(T, self.len);
            errdefer allocator.free(result);
            
            // Copy elements
            @memcpy(result[0..self.len], self.items[0..self.len]);
            
            // Sort using heap sort
            const Context = struct {
                items: []T,
                compare: *const fn (a: T, b: T) Order,
                
                fn lessThan(ctx: @This(), a: usize, b: usize) bool {
                    return ctx.compare(ctx.items[a], ctx.items[b]) == .lt;
                }
                
                fn swap(ctx: @This(), a: usize, b: usize) void {
                    const tmp = ctx.items[a];
                    ctx.items[a] = ctx.items[b];
                    ctx.items[b] = tmp;
                }
            };
            
            std.sort.heap(result, Context{
                .items = result,
                .compare = self.compare,
            }, Context.lessThan, Context.swap);
            
            return result;
        }

        // Internal: sift up from index
        fn siftUp(self: *Self, index: usize) void {
            var i = index;
            while (i > 0) {
                const parent = (i - 1) / 2;
                if (self.compare(self.items[i], self.items[parent]) == .lt) {
                    self.swap(i, parent);
                    i = parent;
                } else {
                    break;
                }
            }
        }

        // Internal: sift down from index
        fn siftDown(self: *Self, index: usize) void {
            var i = index;
            const half = self.len / 2;
            
            while (i < half) {
                var child = 2 * i + 1;
                const right = child + 1;
                
                if (right < self.len and self.compare(self.items[right], self.items[child]) == .lt) {
                    child = right;
                }
                
                if (self.compare(self.items[i], self.items[child]) != .gt) {
                    break;
                }
                
                self.swap(i, child);
                i = child;
            }
        }

        // Internal: swap two elements
        fn swap(self: *Self, a: usize, b: usize) void {
            const tmp = self.items[a];
            self.items[a] = self.items[b];
            self.items[b] = tmp;
        }
    };
}

/// Comparison function for min-heap (smaller values have higher priority)
pub fn minCompare(comptime T: type) fn (a: T, b: T) Order {
    return struct {
        fn compare(a: T, b: T) Order {
            return std.math.order(a, b);
        }
    }.compare;
}

/// Comparison function for max-heap (larger values have higher priority)
pub fn maxCompare(comptime T: type) fn (a: T, b: T) Order {
    return struct {
        fn compare(a: T, b: T) Order {
            return std.math.order(b, a);
        }
    }.compare;
}

/// Comparison for tuples (first element as priority)
pub fn tupleCompare(comptime T: type) fn (a: T, b: T) Order {
    return struct {
        fn compare(a: T, b: T) Order {
            return std.math.order(a[0], b[0]);
        }
    }.compare;
}

// Tests
test "PriorityQueue - basic min-heap operations" {
    const allocator = std.testing.allocator;
    
    var pq = PriorityQueue(i32).init(allocator, minCompare(i32));
    defer pq.deinit();
    
    try std.testing.expect(pq.isEmpty());
    try std.testing.expectEqual(@as(usize, 0), pq.count());
    try std.testing.expectEqual(@as(?i32, null), pq.peek());
    
    try pq.insert(5);
    try pq.insert(3);
    try pq.insert(7);
    try pq.insert(1);
    try pq.insert(9);
    
    try std.testing.expect(!pq.isEmpty());
    try std.testing.expectEqual(@as(usize, 5), pq.count());
    try std.testing.expectEqual(@as(i32, 1), pq.peek().?);
    
    // Should remove in ascending order for min-heap
    try std.testing.expectEqual(@as(i32, 1), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 3), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 5), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 7), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 9), pq.remove().?);
    try std.testing.expect(pq.isEmpty());
}

test "PriorityQueue - basic max-heap operations" {
    const allocator = std.testing.allocator;
    
    var pq = PriorityQueue(i32).init(allocator, maxCompare(i32));
    defer pq.deinit();
    
    try pq.insert(5);
    try pq.insert(3);
    try pq.insert(7);
    try pq.insert(1);
    try pq.insert(9);
    
    // Should remove in descending order for max-heap
    try std.testing.expectEqual(@as(i32, 9), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 7), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 5), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 3), pq.remove().?);
    try std.testing.expectEqual(@as(i32, 1), pq.remove().?);
}

test "PriorityQueue - with capacity" {
    const allocator = std.testing.allocator;
    
    var pq = try PriorityQueue(i32).initCapacity(allocator, 10, minCompare(i32));
    defer pq.deinit();
    
    try std.testing.expectEqual(@as(usize, 10), pq.capacity);
    
    try pq.insert(1);
    try pq.insert(2);
    try std.testing.expectEqual(@as(usize, 2), pq.count());
}

test "PriorityQueue - clear operation" {
    const allocator = std.testing.allocator;
    
    var pq = PriorityQueue(i32).init(allocator, minCompare(i32));
    defer pq.deinit();
    
    try pq.insert(1);
    try pq.insert(2);
    try pq.insert(3);
    
    pq.clear();
    try std.testing.expect(pq.isEmpty());
    try std.testing.expectEqual(@as(usize, 0), pq.count());
}

test "PriorityQueue - dynamic growth" {
    const allocator = std.testing.allocator;
    
    var pq = PriorityQueue(i32).init(allocator, minCompare(i32));
    defer pq.deinit();
    
    // Insert many elements to test growth
    var i: i32 = 100;
    while (i > 0) : (i -= 1) {
        try pq.insert(i);
    }
    
    try std.testing.expectEqual(@as(usize, 100), pq.count());
    
    // Verify order
    i = 1;
    while (i <= 100) : (i += 1) {
        try std.testing.expectEqual(i, pq.remove().?);
    }
}

test "PriorityQueue - tuple priority" {
    const allocator = std.testing.allocator;
    const Tuple = struct { i32, []const u8 };
    
    var pq = PriorityQueue(Tuple).init(allocator, tupleCompare(Tuple));
    defer pq.deinit();
    
    try pq.insert(.{ 3, "low" });
    try pq.insert(.{ 1, "high" });
    try pq.insert(.{ 2, "medium" });
    
    const peeked = pq.peek().?;
    try std.testing.expectEqual(@as(i32, 1), peeked[0]);
    
    const first = pq.remove().?;
    try std.testing.expectEqual(@as(i32, 1), first[0]);
    try std.testing.expectEqualStrings("high", first[1]);
}

test "PriorityQueue - custom struct" {
    const allocator = std.testing.allocator;
    const Task = struct {
        priority: i32,
        name: []const u8,
    };
    
    const compare = struct {
        fn cmp(a: Task, b: Task) Order {
            return std.math.order(a.priority, b.priority);
        }
    }.cmp;
    
    var pq = PriorityQueue(Task).init(allocator, compare);
    defer pq.deinit();
    
    try pq.insert(.{ .priority = 10, .name = "low priority" });
    try pq.insert(.{ .priority = 1, .name = "high priority" });
    try pq.insert(.{ .priority = 5, .name = "medium priority" });
    
    const first = pq.remove().?;
    try std.testing.expectEqual(@as(i32, 1), first.priority);
    try std.testing.expectEqualStrings("high priority", first.name);
}

File: Lua/priority_queue_utils/priority_queue.lua
--- Priority Queue Implementation for Lua
--- A priority queue supports inserting elements with associated priorities
--- and removing the element with the highest (or lowest) priority.
--- Uses a binary heap for efficient O(log n) operations.
--- @class PriorityQueue
--- @field heap table The internal heap array
--- @field compare function The comparison function for ordering (compares priorities directly)
--- @field size number The current number of elements
local PriorityQueue = {}
PriorityQueue.__index = PriorityQueue

--- Creates a new PriorityQueue instance
--- @param is_min_heap boolean|nil If true, creates a min-heap (lowest priority first), 
---                                 if false/nil, creates a max-heap (highest priority first)
--- @return PriorityQueue A new priority queue instance
function PriorityQueue.new(is_min_heap)
    local self = setmetatable({}, PriorityQueue)
    self.heap = {}
    self.size = 0
    
    -- Set comparator based on heap type (compares priority values directly)
    if is_min_heap then
        self.compare = function(a, b) return a < b end
    else
        self.compare = function(a, b) return a > b end
    end
    
    return self
end

--- Creates a max-heap priority queue (highest priority first)
--- @return PriorityQueue A new max-heap priority queue
function PriorityQueue.max_heap()
    return PriorityQueue.new(false)
end

--- Creates a min-heap priority queue (lowest priority first)
--- @return PriorityQueue A new min-heap priority queue
function PriorityQueue.min_heap()
    return PriorityQueue.new(true)
end

--- Creates a priority queue with a custom comparator
--- @param compare function Custom comparison function(a, b) -> boolean
---                         Returns true if priority 'a' should come before 'b'
--- @return PriorityQueue A new priority queue with custom ordering
function PriorityQueue.with_comparator(compare)
    local self = setmetatable({}, PriorityQueue)
    self.heap = {}
    self.size = 0
    self.compare = compare
    return self
end

--- Swaps two elements in the heap
--- @param heap table The heap array
--- @param i number First index
--- @param j number Second index
local function swap(heap, i, j)
    heap[i], heap[j] = heap[j], heap[i]
end

--- Gets the parent index
--- @param i number Current index
--- @return number Parent index
local function parent(i)
    return math.floor(i / 2)
end

--- Gets the left child index
--- @param i number Current index
--- @return number Left child index
local function left_child(i)
    return 2 * i
end

--- Gets the right child index
--- @param i number Current index
--- @return number Right child index
local function right_child(i)
    return 2 * i + 1
end

--- Moves an element up the heap to maintain heap property
--- @param heap table The heap array
--- @param index number Starting index
--- @param compare function The comparison function (compares priorities)
local function heapify_up(heap, index, compare)
    while index > 1 do
        local p = parent(index)
        if compare(heap[index].priority, heap[p].priority) then
            swap(heap, index, p)
            index = p
        else
            break
        end
    end
end

--- Moves an element down the heap to maintain heap property
--- @param heap table The heap array
--- @param index number Starting index
--- @param size number Current heap size
--- @param compare function The comparison function (compares priorities)
local function heapify_down(heap, index, size, compare)
    while true do
        local l = left_child(index)
        local r = right_child(index)
        local extreme = index
        
        if l <= size and compare(heap[l].priority, heap[extreme].priority) then
            extreme = l
        end
        
        if r <= size and compare(heap[r].priority, heap[extreme].priority) then
            extreme = r
        end
        
        if extreme ~= index then
            swap(heap, index, extreme)
            index = extreme
        else
            break
        end
    end
end

--- Inserts an element with a given priority
--- @param value any The value to insert
--- @param priority number The priority of the element
--- @return PriorityQueue self (for method chaining)
function PriorityQueue:push(value, priority)
    self.size = self.size + 1
    self.heap[self.size] = { value = value, priority = priority }
    heapify_up(self.heap, self.size, self.compare)
    return self
end

--- Alias for push
--- @param value any The value to insert
--- @param priority number The priority of the element
--- @return PriorityQueue self
function PriorityQueue:enqueue(value, priority)
    return self:push(value, priority)
end

--- Removes and returns the element with highest priority
--- @return any value, number priority The value and priority of the removed element,
---                                      or nil if queue is empty
function PriorityQueue:pop()
    if self.size == 0 then
        return nil, nil
    end
    
    local root = self.heap[1]
    self.heap[1] = self.heap[self.size]
    self.heap[self.size] = nil
    self.size = self.size - 1
    
    if self.size > 0 then
        heapify_down(self.heap, 1, self.size, self.compare)
    end
    
    return root.value, root.priority
end

--- Alias for pop
--- @return any value, number priority
function PriorityQueue:dequeue()
    return self:pop()
end

--- Returns the element with highest priority without removing it
--- @return any value, number priority The value and priority of the top element,
---                                      or nil if queue is empty
function PriorityQueue:peek()
    if self.size == 0 then
        return nil, nil
    end
    return self.heap[1].value, self.heap[1].priority
end

--- Checks if the queue is empty
--- @return boolean True if empty, false otherwise
function PriorityQueue:is_empty()
    return self.size == 0
end

--- Returns the number of elements in the queue
--- @return number The size of the queue
function PriorityQueue:len()
    return self.size
end

--- Clears all elements from the queue
function PriorityQueue:clear()
    self.heap = {}
    self.size = 0
end

--- Converts the queue to a sorted array (without modifying the queue)
--- @return table Array of {value, priority} pairs in priority order
function PriorityQueue:to_sorted_array()
    local result = {}
    -- Collect all elements
    for i = 1, self.size do
        table.insert(result, { value = self.heap[i].value, priority = self.heap[i].priority })
    end
    
    -- Sort using the same comparison logic
    table.sort(result, function(a, b)
        return self.compare(a.priority, b.priority)
    end)
    
    return result
end

--- Updates the priority of an element
--- @param value any The value to find
--- @param new_priority number The new priority
--- @param equals function|nil Optional equality function(a, b) -> boolean
--- @return boolean True if element was found and updated, false otherwise
function PriorityQueue:update_priority(value, new_priority, equals)
    equals = equals or function(a, b) return a == b end
    
    for i = 1, self.size do
        if equals(self.heap[i].value, value) then
            local old_priority = self.heap[i].priority
            self.heap[i].priority = new_priority
            
            -- Determine if we need to heapify up or down
            if self.compare(new_priority, old_priority) then
                heapify_up(self.heap, i, self.compare)
            else
                heapify_down(self.heap, i, self.size, self.compare)
            end
            
            return true
        end
    end
    
    return false
end

--- Checks if a value exists in the queue
--- @param value any The value to find
--- @param equals function|nil Optional equality function
--- @return boolean True if found, false otherwise
function PriorityQueue:contains(value, equals)
    equals = equals or function(a, b) return a == b end
    
    for i = 1, self.size do
        if equals(self.heap[i].value, value) then
            return true
        end
    end
    
    return false
end

--- Removes a specific value from the queue
--- @param value any The value to remove
--- @param equals function|nil Optional equality function
--- @return boolean True if removed, false if not found
function PriorityQueue:remove(value, equals)
    equals = equals or function(a, b) return a == b end
    
    for i = 1, self.size do
        if equals(self.heap[i].value, value) then
            -- Replace with last element
            self.heap[i] = self.heap[self.size]
            self.heap[self.size] = nil
            self.size = self.size - 1
            
            if i <= self.size then
                -- Try both heapify directions
                heapify_up(self.heap, i, self.compare)
                heapify_down(self.heap, i, self.size, self.compare)
            end
            
            return true
        end
    end
    
    return false
end

--- Merges another priority queue into this one
--- @param other PriorityQueue Another priority queue to merge
--- @return PriorityQueue self (for method chaining)
function PriorityQueue:merge(other)
    for i = 1, other.size do
        self:push(other.heap[i].value, other.heap[i].priority)
    end
    return self
end

--- Returns the internal heap as an array (for debugging)
--- @return table The internal heap array
function PriorityQueue:debug_heap()
    local result = {}
    for i = 1, self.size do
        table.insert(result, { value = self.heap[i].value, priority = self.heap[i].priority })
    end
    return result
end

return PriorityQueue